File: singleton.go
package main

// 利用Golang的语言特性，使用sync.Once以只执行一次实例创建

import (
	"fmt"
	"sync"
)

var instance *Singleton
var once sync.Once

type Singleton struct {
	n int
}

func GetInstance() *Singleton {
	once.Do(func() {
		instance = &Singleton{}
	})
	return instance
}

func main() {
	a := GetInstance()
	fmt.Println(a)

	a.n = 123

	b := GetInstance()
	fmt.Println(b)
}

File: singleton.cpp
/*
    剑指offer 1, 单例模式
*/
#include <iostream>
#include <pthread.h>
using namespace std;

class Singleton
{
    public:
        int value;
        static Singleton& getInstance(int v)
        {
            static Singleton instance;
            instance.value = v;
            return instance;
        }
};

class Singleton2
{
    private:
        Singleton2(){}
        static Singleton2* p;
    public:
        static pthread_mutex_t mutex;
        static Singleton2* getInstance();
};
pthread_mutex_t Singleton2::mutex;
Singleton2* Singleton2::p = NULL;
Singleton2* Singleton2::getInstance()
{
    if (p != NULL) return p;
    pthread_mutex_lock(&mutex);
    if (p == NULL) p = new Singleton2();
    pthread_mutex_unlock(&mutex);
    return p;
}

int main()
{
    // Singleton a = Singleton::getInstance(1);
    // Singleton b = Singleton::getInstance(2);
    // cout << &a << a.value << endl;
    // cout << &b << b.value << endl;

    pthread_mutex_init(&Singleton2::mutex, NULL);

    Singleton2* a = Singleton2::getInstance();
    Singleton2* b = Singleton2::getInstance();
    cout << a << endl;
    cout << b << endl;
}
